### ISSUE TITLE
Should not emit EmptyArrayAccess on variables that are references

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#438]
Expected: no warning
Observed:  EmptyArrayAccess - src/example.php:6:20 - Cannot access value on empty array variable $x

One possible solution is to give up on tracking certain inferences about references?

```php
<?php
function testRef() : array {
    $result = [];
    foreach ([1, 2, 1] as $v) {
        $x = &$result;
        if (!isset($x[$v])) {
            $x[$v] = 0;
        }
        $x[$v] ++;
    }
    return $result;
}
```

Other variant: Closure passing reference to array_map

**Comment by muglug** [comment]
Would be fixed by https://github.com/vimeo/psalm/issues/103
